reference/behavior_test/establishing_decorator_pattern.rst

```diff
@@ -0,0 +1,367 @@
+
+        $this->wait();
+
+        return $page;
+    }
+
+Selenium will know which URL to go to thanks to `the protected $path attribute
+<https://github.com/akeneo/pim-community-dev/blob/master/features/Context/Page/Role/Edit.php#L19>`_
+containing in *Context\\Page\\Role\\Edit*.
+
+You may notice that this Gherkin method still uses the old implementation of Behat in Akeneo PIM.
+
+Call Decorator Method
+*********************
+
+*And I grant rights to group Products*
+
+All our new contexts are in the namespace *\\Pim\\Behat\\Context\\Domain* which is classified by domain. Each domain is
+linked to a page in Akeneo PIM. That's why, in this namespace, you will find the same architecture as in our main menu.
+This organization is very helpful to quickly find the context you are looking for.
+
+We are working on permissions that are localized in the system main menu. So let's see the
+`\\Pim\\Behat\\Context\\Domain\\System\\PermissionsContext
+<https://github.com/akeneo/pim-community-dev/blob/master/features/Pim/Behat/Context/Domain/System/PermissionsContext.php>`_.
+
+.. image:: images/PermissionsContext.png
+
+.. code-block:: php
+
+    // \Pim\Behat\Context\Domain\System\PermissionsContext
+    /**
+     * @param string $action (grant|remove)
+     * @param string $group
+     *
+     * Gherkin will match this regexp
+     * @When /^I (grant|revoke) rights to groups? (.*)$/
+     */
+    public function iSetRightsToACLGroups($action, $groups) // $action = 'grant', $group = 'Products'
+    {
+        foreach ($this->listToArray($groups) as $group) {
+            // In the previous step we put the current page to "Role edit"
+            // Returns an instance of Context\Page\Role\Edit
+            $page = $this->getCurrentPage();
+
+            // Returns an instance of Pim\Behat\Decorator\Permission\PermissionDecorator
+            $permissionElement = $page->getElement('Permission');
+
+            switch ($action) {
+                case 'grant':
+                    $method = 'grantGroup';
+                    break;
+                case 'revoke':
+                    $method = 'revokeGroup';
+                    break;
+                default:
+                    throw new \InvalidArgumentException(sprintf('Action "%s" does not exist.', $action));
+                    break;
+            }
+
+            foreach ($this->listToArray($groups) as $group) {
+                // We call the method grantGroup() from the PermissionDecorator
+                $permissionElement->$method($group);
+            }
+        }
+    }
+
+Pages
+_____
+
+The current page is `Context\\Page\\Role\\Edit
+<https://github.com/akeneo/pim-community-dev/blob/master/features/Context/Page/Role/Edit.php>`_.
+
+When you call *Pim\\Behat\\Context\\NavigationContext->getCurrentPage()* the method will match "Role edit" to
+*return new Context\\Page\\Role\\Edit()*. That's why all our pages are classified in a folder with the Entity name and
```

Comment:
This sentence is too long and you should split it:
That's why all of our pages are located in a folder named after the name of the Entity. In this folder, each action has its dedicated page such as `Index.php`, [...] for example.